cache/slice.go

```diff
@@ -0,0 +1,35 @@
+package cache
+
+import "io"
+
+type SliceReadWriter struct {
+	slice    [][]byte
+	position uint64
+}
+
+func (s *SliceReadWriter) Width() uint64 {
+	return uint64(len(s.slice))
+}
+
+func (s *SliceReadWriter) Seek(index uint64) error {
+	if index >= uint64(len(s.slice)) {
+		return io.EOF
+	}
+	s.position = index
+	return nil
+}
+
+func (s *SliceReadWriter) ReadNext() ([]byte, error) {
+	if s.position >= uint64(len(s.slice)) {
+		return nil, io.EOF
+	}
+	value := make([]byte, NodeSize)
+	copy(value, s.slice[s.position])
+	s.position++
+	return value, nil
+}
+
+func (s *SliceReadWriter) Write(p []byte) (n int, err error) {
```

Comment:
This is a little confusing, I agree. Not sure how to fix it though.

`Write` is the `Writer` part, while the other methods are the `Reader` part. They are independent and `position` is only relevant to the `Reader` part. `Write` is append-only.

I'm not sure what's the best way to keep this behaviour and also make it clear to an uninformed user. If I change the function name to `Append` that would break the `io.Writer` interface implementation, but maybe that's the best solution. WDYT?

Reply:
I wouldn't be the first to have a combined read-writer that actually work independently. See go's [`bufio.ReadWriter`](https://golang.org/pkg/bufio/#ReadWriter) -- its reader and writer work in isolation and don't share the position with each other.

I'm leaving as-is, unless you tell me otherwise.

Reply:
Rename `Write` to `Append` and add a comment next to `Seek` saying that it only effects the reader